### DISCUSSION TITLE
firebase.messaging().getToken() returns null

### DISCUSSION THREAD

**Discussion by toytonic** [invertase/react-native-firebase#4926] [Q&A]
Hi there,
before filing a bug I would like to ask for any feedback if you experienced similar behavior. We use Push Notifications extensively and experienced that sometimes:

`const token = await messaging().getToken()` returns `null`

I never saw it locally but we definitely see it in our production logs, mostly ios. Any thoughts?

Thanks,
Christian

**Comment by mikehardy** [comment]
If you search in the issues list I believe others have reported this. Hard to say if it was a firebase-ios-sdk issue or not, if it was whether it was corrected in current versions or not, or project-specific issue or not. I do know that getting tokens is an asynchronous process based on the network so I'd check the documentation of the underlying firebase-ios-sdk very carefully and see if the expectations are reasonable. May be that null is an acceptable return value under certain conditions.

**Reply by toytonic** [reply]
Alright, thanks for your quick reply. I will try to find the issues you mentioned and take a look at the sdk, although my native ios skills are limited ;) Our messaging setup is as simple as possible. We fetch the token on startup and have an update handler onTokenRefresh. So I can almost 100% exclude any project-specific issue.

If `null` is a valid value of `await messaging().getToken()` we should reflect that in the docs and typescript definitions.

**Reply by toytonic** [reply]
So far so good. I took a look at the native implementations and couldn't find anything that indicates that an empty token is a valid return value. One thing that I did notice though is that the sdk module `FirebaseInstanceId` being used for retrieving the token  is deprecated.

<img width="915" alt="Bildschirmfoto 2021-02-19 um 11 41 20" src="https://user-images.githubusercontent.com/248213/108497353-99308500-72ab-11eb-81e4-0999680ff2ae.png">

I will add even more logging to our app to see if maybe `await messaging.getToken()` just never has been resolved until we need it.

**Reply by mikehardy** [reply]
Indeed Instance ID APIs are deprecated - we have not moved to the new Installations module and its APIs yet, however I believe internally they already have so it is more about bending API usage to fit current implementation vs following different code paths internal to the firebase native SDKs. I will warn that statement is supposition but it is "to the best of my knowledge".

I will also say token handling is something that may very well have some corner cases with errors here - I won't be surprised if react-native-firebase is missing a case or handling a case poorly and I'll be keen to merge a PR if so, but it is such a vital thing for apps that use this module that I'd like to see the full reasoning and links to docs if so
